src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs

```diff
@@ -275,108 +275,174 @@ public static void ValidateHeaderNameCharacters(string headerCharacters)
-                                }
+                            offset += sizeof(ulong) / 2;
+                            byteValue = byteValue.Slice(sizeof(ulong));
 
-                                if (ch == tokenEnd)
-                                {
-                                    connectionOptions |= ConnectionOptions.Upgrade;
-                                }
+                            if (ReadLowerCaseUInt32(byteValue) == upgradeEnd)
+                            {
+                                offset += sizeof(uint) / 2;
+                                potentialConnectionOptions = ConnectionOptions.Upgrade;
                             }
                         }
-                        else if (tokenLength >= 5 && (*ch | 0x20) == 'c')
+                    }
+                    else if (c == 'c' && byteValue.Length >= sizeof(ulong))
+                    {
+                        if (ReadLowerCaseUInt64(byteValue) == closeEnd)
                         {
-                            if ((*++ch | 0x20) == 'l' &&
-                                (*++ch | 0x20) == 'o' &&
-                                (*++ch | 0x20) == 's' &&
-                                (*++ch | 0x20) == 'e')
-                            {
-                                ch++;
-                                while (ch < tokenEnd && *ch == ' ')
-                                {
-                                    ch++;
-                                }
+                            offset += sizeof(ulong) / 2;
+                            potentialConnectionOptions = ConnectionOptions.Close;
+                        }
+                    }
 
-                                if (ch == tokenEnd)
-                                {
-                                    connectionOptions |= ConnectionOptions.Close;
-                                }
-                            }
+                    if ((uint)offset >= (uint)value.Length)
+                    {
+                        // Consumed enitre string, move to next string.
+                        connectionOptions |= potentialConnectionOptions;
+                        break;
+                    }
+                    else
+                    {
+                        value = value.Slice(offset);
+                    }
+
+                    for (offset = 0; offset < value.Length; offset++)
+                    {
+                        c = value[offset];
+                        if (c == ' ')
+                        {
+                            continue;
                         }
+                        if (c == ',')
+                        {
+                            break;
+                        }
+                        else
+                        {
+                            // Value contains extra chars; this is not the matched one.
+                            potentialConnectionOptions = ConnectionOptions.None;
+                            continue;
```

Comment:
Nit: this `continue;` can just be removed right? It would be easier to read if this was just `if (c == ',') break; if (c != ' ') potentialConnectionOptions = ConnectionOptions.None;`. With the appropriate braces and newlines ofc.

Reply:
Done